### PR TITLE
DT-635 audit proxy user

### PR DESCRIPTION
**Pull request by mikehalmamoj**
Improve auditing around updates from client credential JWTs.
* Full audit details are written to the DB for Nomis users
* Other recognised (e.g. AUth, Delius) users will have their username appear in the audit details.
* Anonymous users will have the IP address and URL written to the audit details

Note that non-Nomis users may receive a 404 if the update only applies to Nomis users

### CODE REVIEW

File: src/main/java/net/syscon/elite/web/filter/UserMdcFilter.java [hunk 1]
```diff
@@ -49,17 +49,18 @@ public void doFilter(final ServletRequest request, final ServletResponse respons
             if (currentUsername != null) {
                 MDC.put(USER_ID_HEADER, currentUsername);
                 telemetryClient.getContext().getUser().setId(currentUsername);
-                telemetryClient.getContext().getComponent().setVersion(getVersion());
-                final var ip = IpAddressHelper.retrieveIpFromRemoteAddr((HttpServletRequest) request);
-                MDC.put(IP_ADDRESS, ip);
-                telemetryClient.getContext().getLocation().setIp(ip);
             }
+            telemetryClient.getContext().getComponent().setVersion(getVersion());
+            final var ip = IpAddressHelper.retrieveIpFromRemoteAddr((HttpServletRequest) request);
```
Comment (mikehalmamoj) [inline src/main/java/net/syscon/elite/web/filter/UserMdcFilter.java]: Capture the IP address even for anonymous users

File: src/main/java/net/syscon/elite/web/config/AuthAwareAuthenticationToken.java [hunk 2]
```diff
@@ -9,16 +10,12 @@
 
 @Getter
 public class AuthAwareAuthenticationToken extends JwtAuthenticationToken {
-    private final String authSource;
+    private final AuthSource authSource;
     private final Object principal;
 
     public AuthAwareAuthenticationToken(final Jwt jwt, final Object principal, final String authSource, final Collection<? extends GrantedAuthority> authorities) {
         super(jwt, authorities);
-        this.authSource = authSource != null ? authSource : "nomis";
```
Comment (mikehalmamoj) [inline src/main/java/net/syscon/elite/web/config/AuthAwareAuthenticationToken.java]: We no longer default to a Nomis user as this is misleading.

File: src/main/java/net/syscon/elite/aop/connectionproxy/HsqlConnectionAspect.java [hunk 3]
```diff
@@ -19,7 +21,8 @@ public HsqlConnectionAspect(final AuthenticationFacade authenticationFacade) {
 
     @Override
     protected Connection openProxySessionIfIdentifiedAuthentication(final Connection pooledConnection) throws SQLException {
-        if (authenticationFacade.isIdentifiedAuthentication()) {
+        final var proxyUserAuthSource = authenticationFacade.getProxyUserAuthenticationSource();
+        if (proxyUserAuthSource.equals(NOMIS)) {
```
Comment (petergphillips) [inline src/main/java/net/syscon/elite/aop/connectionproxy/HsqlConnectionAspect.java]: don't need `.equals` for enums

File: src/main/java/net/syscon/elite/aop/connectionproxy/OracleConnectionAspect.java [hunk 4]
```diff
@@ -34,12 +35,13 @@ public OracleConnectionAspect(
 
     @Override
     protected Connection openProxySessionIfIdentifiedAuthentication(final Connection pooledConnection) throws SQLException {
-        if (authenticationFacade.isIdentifiedAuthentication()) {
-            log.debug("Configuring Oracle Proxy Session {}", pooledConnection);
+        final var proxyUserAuthSource = authenticationFacade.getProxyUserAuthenticationSource();
+        if (proxyUserAuthSource.equals(NOMIS)) {
```
Comment (petergphillips) [inline src/main/java/net/syscon/elite/aop/connectionproxy/OracleConnectionAspect.java]: don't need `.equals` for enums

File: src/main/java/net/syscon/elite/aop/connectionproxy/OracleConnectionAspect.java [hunk 5]
```diff
@@ -48,15 +50,19 @@ private Connection openAndConfigureProxySessionForConnection(final Connection po
...
-
         roleConfigurer.setRoleForConnection(oracleConnection);
 
-        setContext(wrappedConnection);
+        configureConnection(wrappedConnection);
 
         return wrappedConnection;
     }
 
+    private Connection configureConnection(Connection pooledConnection) throws SQLException {
```
Comment (petergphillips) [inline src/main/java/net/syscon/elite/aop/connectionproxy/OracleConnectionAspect.java]: missing `final`

File: src/main/java/net/syscon/elite/web/config/AuthAwareAuthenticationToken.java (hunk 2 above)
Comment (petergphillips) [reply src/main/java/net/syscon/elite/web/config/AuthAwareAuthenticationToken.java]: :+1:

File: src/main/java/net/syscon/elite/security/AuthSource.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    static {
+        Arrays.stream(AuthSource.values()).forEach(authSource -> byName.put(authSource.name, authSource));
+    }
+
+    AuthSource(final String name) {
+        this.name = name;
+    }
+
+    public static AuthSource fromName(final String name) {
+        return byName.getOrDefault(name, NONE);
```
Comment (matt-crowson-moj) [inline src/main/java/net/syscon/elite/security/AuthSource.java]: Perhaps make case insensitive, not sure really, meh!

File: src/main/java/net/syscon/elite/security/AuthenticationFacade.java [hunk 7]
```diff
@@ -44,14 +44,15 @@ public String getCurrentUsername() {
...
-        return StringUtils.isNotBlank(MDC.get(PROXY_USER))
-                && Optional.ofNullable(auth).
+        return Optional.ofNullable(auth).
+                filter(a -> StringUtils.isNotBlank(MDC.get(PROXY_USER))).
                 filter(AuthAwareAuthenticationToken.class::isInstance).
                 map(AuthAwareAuthenticationToken.class::cast).
                 filter(AbstractAuthenticationToken::isAuthenticated).
-                map(AuthAwareAuthenticationToken::isNomisSource).orElse(Boolean.FALSE);
+                map(AuthAwareAuthenticationToken::getAuthSource).
+                orElse(AuthSource.NONE);
```
Comment (matt-crowson-moj) [inline src/main/java/net/syscon/elite/security/AuthenticationFacade.java]: Not really an issue.  But prefer the supplier (lazy) version of this.  `.orElseGet(() -> ...);`.  I have been burnt in the past with one.

**Pull request merged**